ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,456 @@
+        Self { block: range }
+    }
+}
+
+pub struct VirtAddrAllocator {
+    freelist: BTreeMap<Vaddr, KvaFreeNode>,
+}
+
+impl VirtAddrAllocator {
+    fn new(range: Range<Vaddr>) -> Self {
+        let mut freelist: BTreeMap<Vaddr, KvaFreeNode> = BTreeMap::new();
+        freelist.insert(range.start, KvaFreeNode::new(range));
+        Self { freelist }
+    }
+
+    /// Allocate a kernel virtual area.
+    ///
+    /// This is currently implemented with a simple FIRST-FIT algorithm.
+    fn alloc(&mut self, size: usize) -> Result<Range<Vaddr>> {
+        let mut allocate_range = None;
+        let mut to_remove = None;
+
+        for (key, value) in self.freelist.iter() {
+            if value.block.end - value.block.start >= size {
+                allocate_range = Some((value.block.end - size)..value.block.end);
+                to_remove = Some(*key);
+                break;
+            }
+        }
+
+        if let Some(key) = to_remove {
+            if let Some(freenode) = self.freelist.get_mut(&key) {
+                if freenode.block.end - size == freenode.block.start {
+                    self.freelist.remove(&key);
+                } else {
+                    freenode.block.end -= size;
+                }
+            }
+        }
+
+        if let Some(range) = allocate_range {
+            Ok(range)
+        } else {
+            Err(Error::KvaAllocError)
+        }
+    }
+
+    /// Free a kernel virtual area.
+    fn free(&mut self, range: Range<Vaddr>) {
+        // 1. get the previous free block, check if we can merge this block with the free one
+        //     - if contiguous, merge this area with the free block.
+        //     - if not contiguous, create a new free block, insert it into the list.
+        let mut free_range = range.clone();
+
+        if let Some((prev_va, prev_node)) = self
+            .freelist
+            .upper_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_prev()
+        {
+            if prev_node.block.end == free_range.start {
+                let prev_va = *prev_va;
+                free_range.start = prev_node.block.start;
+                self.freelist.remove(&prev_va);
+            }
+        }
+        self.freelist
+            .insert(free_range.start, KvaFreeNode::new(free_range.clone()));
+
+        // 2. check if we can merge the current block with the next block, if we can, do so.
+        if let Some((next_va, next_node)) = self
+            .freelist
+            .lower_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_next()
+        {
+            if free_range.end == next_node.block.start {
+                let next_va = *next_va;
+                free_range.end = next_node.block.end;
+                self.freelist.remove(&next_va);
+                self.freelist.get_mut(&free_range.start).unwrap().block.end = free_range.end;
+            }
+        }
+    }
+}
+
+lazy_static! {
```

Comment:
It would be better to avoid `lazy_static` in our kernel code because it has racing issues. The underlying `spin_no_std` implementation doesn't take our preemption mechanism into account. It is also generally not encouraged for std environments either: https://github.com/rust-lang-nursery/lazy-static.rs/issues/214.

A recommended approach is to make `VirtAddrAllocator::new` const (it is ok since `BTreeSet::new` is const). Then it could do static init.

Otherwise, if the allocator in the future indeed cannot be const, please follow the recommended approach in the lazy static README and use `SpinLock<Option>` to replace `OnceLock`